### PR TITLE
Add support for LayerNorm before residual connection in transformer

### PR DESCRIPTION
**Pull request by jarednielsen**
Recent work has shown that placing the layer norm before the residual connection in a transformer leads to better gradient propagation and more stable training. For example, [Learning Deep Transformer Models for Machine Translation](https://arxiv.org/abs/1906.01787) shows that when stacking more layers, the Pre-LN transformer performs better than the Post-LN transformer. A more detailed discussion is found in [On Layer Normalization in the Transformer Architecture](https://openreview.net/pdf?id=B1x8anVFPr), which aids understanding despite being rejected from ICLR 2020 for lack of novelty.

_In my own experiments pretraining from scratch with TFAlbert, I have found pre-layer normalization to have greater training stability and improved performance than the default post-layer normalization._

Screen Shot 2020-04-23 at 11 37 36 AM Screen Shot 2020-04-23 at 11 41 05 AM

**This is not a final PR, just opening the discussion. Some caveats:**
- If approved, I can extend this to all transformer models, not just ALBERT.
- Loading an existing pretrained model with `config.pre_layer_norm = True` leads to poor performance. A separate pretrained model would need to be made available with this configuration option. I would be happy to do so for ALBERT, but this would be outside my pay grade for all the models. Since the current focus of the repo is on finetuning rather than pretraining, I understand if this isn't a complexity tradeoff you want to make.

What do you think about adding this configuration option?

### CODE REVIEW

**Comment by Evpok** [comment]
Because I just stumbled upon it: a counterpoint in [Understanding the Difficulty of Training Transformers](https://arxiv.org/abs/2004.08249), where the author founds that while this tends to stabilize the convergence, it also results in worse performances for the runs that do converge.

(Still, nice PR, I was toying with the idea of doing it too :-)

**Comment by di-he** [comment]
> Because I just stumbled upon it: a counterpoint in [Understanding the Difficulty of Training Transformers](https://arxiv.org/abs/2004.08249), where the author founds that while this tends to stabilize the convergence, it also results in worse performances for the runs that do converge.
> 
> (Still, nice PR, I was toying with the idea of doing it too :-)

Hi, author here (On Layer Normalization in the Transformer Architecture),

For Pre-LN, a slightly larger dropout rate is usually needed for median scale tasks, such as translation. This is not required for BERT. Pre-LN has larger expressiveness power than Post-LN and when the data scale is small, it is more likely to overfit. 

When you use a larger dropout in translation tasks, Pre-LN is no worse than Post-LN. [There are three 'dropout' places in Transformer, For translation tasks, we apply 0.1 to all the three places which make Pre-LN is faster to converge and achieves better/competitive performance]. 

You can reproduce the experiments in our paper.

**Comment by Evpok** [comment]
Oh, thanks for chiming in, I must read the paper again then!

**Comment by jarednielsen** [comment]
Bumping this discussion. Any thoughts @LysandreJik @julien-c ?

**Comment by jarednielsen** [comment]
Not stale

**Comment by di-he** [comment]
> Not stale

I think the pre-ln version should be added to the branch. As far as I know, there are more and more works that are developed based on this architecture. BTW, although our paper is rejected by ICLR, it is accepted by ICML:)

**Comment by jarednielsen** [comment]
Bumping this, any thoughts @thomwolf ?

File: src/transformers/modeling_tf_albert.py [hunk 1]
```diff
@@ -324,16 +330,23 @@ def __init__(self, config, **kwargs):
         )
         self.dropout = tf.keras.layers.Dropout(config.hidden_dropout_prob)
 
+        self.pre_layer_norm = config.pre_layer_norm
+
     def call(self, inputs, training=False):
         hidden_states, attention_mask, head_mask = inputs
 
+        if self.pre_layer_norm:
```
Comment (likicode) [inline src/transformers/modeling_tf_albert.py:338]: I think this line (pre_layer_norm) should be moved after the call of self.attention (Line 340). There is already a pre_layer_norm process in TFAlbertAttention. According to the paper (ON LAYER NORMALIZATION IN THE TRANSFORMER ARCHITECTURE), there should be another pre_layer_norm before FFN sub-layer. Therefore I assume Line 338 should be followed by Line 340.

**Comment by jarednielsen** [comment]
Not stale
